File: clients/lobby-sdk/ts/src/codec.ts
// Wire-format codec for the Silencer lobby protocol.
// See ../../../shared/lobby-protocol/protocol.md.

import {
    type AgencyStats,
    type AuthResult,
    type ChatMessage,
    emptyAgency,
    type LobbyGame,
    type MatchStats,
    type MotdChunk,
    type NewGameEvent,
    Op,
    Platform,
    type PresenceUpdate,
    SecurityLevel,
    GameStatus,
    type UserInfo,
    type VersionResult,
    type WeaponStats,
} from "./types.ts";

export class CodecError extends Error {
    override name = "CodecError";
}

export const MAX_FRAME_PAYLOAD = 255;

// ---- Reader / Writer ----------------------------------------------------

export class Reader {
    private off = 0;
    constructor(public readonly bytes: Uint8Array) {}

    get remaining(): number {
        return this.bytes.length - this.off;
    }
    get offset(): number {
        return this.off;
    }

    u8(): number {
        if (this.off >= this.bytes.length) throw new CodecError("u8: short read");
        return this.bytes[this.off++]!;
    }

    u16Le(): number {
        if (this.off + 2 > this.bytes.length) throw new CodecError("u16: short read");
        const v = this.bytes[this.off]! | (this.bytes[this.off + 1]! << 8);
        this.off += 2;
        return v;
    }

    u32Le(): number {
        if (this.off + 4 > this.bytes.length) throw new CodecError("u32: short read");
        const v =
            this.bytes[this.off]! |
            (this.bytes[this.off + 1]! << 8) |
            (this.bytes[this.off + 2]! << 16) |
            (this.bytes[this.off + 3]! << 24);
        this.off += 4;
        // Force unsigned interpretation.
        return v >>> 0;
    }

    bytesN(n: number): Uint8Array {
        if (this.off + n > this.bytes.length) throw new CodecError("bytes: short read");
        const out = this.bytes.slice(this.off, this.off + n);
        this.off += n;
        return out;
    }

    cstr(maxLen: number): string {
        const limit = Math.min(this.bytes.length, this.off + maxLen);
        let end = this.off;
        while (end < limit && this.bytes[end] !== 0) end++;
        if (end >= limit) throw new CodecError("cstr: unterminated");
        const s = utf8Decode(this.bytes.subarray(this.off, end));
        this.off = end + 1;
        return s;
    }

    lenstr(): string {
        const n = this.u8();
        if (this.off + n > this.bytes.length) throw new CodecError("lenstr: short read");
        const s = utf8Decode(this.bytes.subarray(this.off, this.off + n));
        this.off += n;
        return s;
    }
}

export class Writer {
    private chunks: number[] = [];

    u8(v: number): void {
        this.chunks.push(v & 0xff);
    }
    u16Le(v: number): void {
        this.chunks.push(v & 0xff, (v >>> 8) & 0xff);
    }
    u32Le(v: number): void {
        this.chunks.push(v & 0xff, (v >>> 8) & 0xff, (v >>> 16) & 0xff, (v >>> 24) & 0xff);
    }
    write(src: Uint8Array): void {
        for (let i = 0; i < src.length; i++) this.chunks.push(src[i]!);
    }
    cstr(s: string): void {
        const b = utf8Encode(s);
        for (let i = 0; i < b.length; i++) this.chunks.push(b[i]!);
        this.chunks.push(0);
    }
    lenstr(s: string): void {
        const b = utf8Encode(s);
        const n = Math.min(b.length, 255);
        this.chunks.push(n);
        for (let i = 0; i < n; i++) this.chunks.push(b[i]!);
    }
    bytes(): Uint8Array {
        return new Uint8Array(this.chunks);
    }
}

// UTF-8 helpers (Bun has TextEncoder/TextDecoder globally).
const _enc = new TextEncoder();
const _dec = new TextDecoder("utf-8", { fatal: false });
const utf8Encode = (s: string) => _enc.encode(s);
const utf8Decode = (b: Uint8Array) => _dec.decode(b);

// ---- Framing ------------------------------------------------------------

export function frameEncode(payload: Uint8Array): Uint8Array {
    if (payload.length === 0 || payload.length > MAX_FRAME_PAYLOAD) {
        throw new CodecError("frameEncode: bad payload size");
    }
    const out = new Uint8Array(1 + payload.length);
    out[0] = payload.length;
    out.set(payload, 1);
    return out;
}

// Returns the consumed payload (opcode + body) and how many bytes of the
// input were consumed, or null if a full frame isn't yet available.
export function frameTryDecode(buf: Uint8Array): { payload: Uint8Array; consumed: number } | null {
    if (buf.length < 1) return null;
    const n = buf[0]!;
    if (n === 0) throw new CodecError("frameTryDecode: zero-length frame");
    if (buf.length < 1 + n) return null;
    return { payload: buf.slice(1, 1 + n), consumed: 1 + n };
}

// ---- LobbyGame ----------------------------------------------------------

export function encodeLobbyGame(w: Writer, g: LobbyGame): void {
    w.u32Le(g.id);
    w.u32Le(g.accountId);
    w.lenstr(g.name);
    w.lenstr(g.password);
    w.lenstr(g.hostname);
    w.lenstr(g.mapName);
    if (g.mapHash.length !== 20) throw new CodecError("mapHash must be 20 bytes");
    w.write(g.mapHash);
    w.u8(g.players);
    w.u8(g.state);
    w.u8(g.securityLevel);
    w.u8(g.minLevel);
    w.u8(g.maxLevel);
    w.u8(g.maxPlayers);
    w.u8(g.maxTeams);
    w.u8(g.extra);
    w.u16Le(g.port);
}

export function decodeLobbyGame(r: Reader): LobbyGame {
    const id = r.u32Le();
    const accountId = r.u32Le();
    const name = r.lenstr();
    const password = r.lenstr();
    const hostname = r.lenstr();
    const mapName = r.lenstr();
    const mapHash = r.bytesN(20);
    const players = r.u8();
    const state = r.u8();
    const securityLevel = r.u8() as SecurityLevel;
    const minLevel = r.u8();
    const maxLevel = r.u8();
    const maxPlayers = r.u8();
    const maxTeams = r.u8();
    const extra = r.u8();
    const port = r.u16Le();
    return {
        id, accountId, name, password, hostname, mapName, mapHash,
        players, state, securityLevel, minLevel, maxLevel,
        maxPlayers, maxTeams, extra, port,
    };
}

// ---- Per-opcode encoders ------------------------------------------------

export function encodeVersionRequest(version: string, platform: Platform): Uint8Array {
    const w = new Writer();
    w.u8(Op.Version);
    w.cstr(version);
    w.u8(platform);
    return w.bytes();
}

export function encodeAuthRequest(username: string, passwordSha1: Uint8Array): Uint8Array {
    if (passwordSha1.length !== 20) throw new CodecError("passwordSha1 must be 20 bytes");
    const w = new Writer();
    w.u8(Op.Auth);
    w.cstr(username);
    w.write(passwordSha1);
    return w.bytes();
}

export function encodeChat(channel: string, message: string): Uint8Array {
    const w = new Writer();
    w.u8(Op.Chat);
    w.cstr(channel);
    w.cstr(message);
    return w.bytes();
}

export function encodeJoinChannel(currentChannel: string, newChannel: string): Uint8Array {
    return encodeChat(currentChannel, "/join " + newChannel);
}

export function encodeNewGame(g: LobbyGame): Uint8Array {
    const w = new Writer();
    w.u8(Op.NewGame);
    encodeLobbyGame(w, g);
    return w.bytes();
}

export function encodeUserInfoRequest(accountId: number): Uint8Array {
    const w = new Writer();
    w.u8(Op.UserInfo);
    w.u32Le(accountId);
    return w.bytes();
}

export function encodePingAck(): Uint8Array {
    return new Uint8Array([Op.Ping, 1]);
}

export function encodeUpgradeStat(agencyIdx: number, statId: number): Uint8Array {
    return new Uint8Array([Op.UpgradeStat, agencyIdx & 0xff, statId & 0xff]);
}

export function encodeSetGame(gameId: number, status: GameStatus): Uint8Array {
    const w = new Writer();
    w.u8(Op.SetGame);
    w.u32Le(gameId);
    w.u8(status);
    return w.bytes();
}

export function encodeRegisterStats(
    gameId: number,
    teamNumber: number,
    accountId: number,
    statsAgency: number,
    won: boolean,
    xp: number,
    s: MatchStats,
): Uint8Array {
    const w = new Writer();
    w.u8(Op.RegisterStats);
    w.u32Le(gameId);
    w.u8(teamNumber);
    w.u32Le(accountId);
    w.u8(statsAgency);
    w.u8(won ? 1 : 0);
    w.u32Le(xp);
    for (let i = 0; i < 4; i++) {
        const wp = s.weapons[i]!;
        w.u32Le(wp.fires);
        w.u32Le(wp.hits);
        w.u32Le(wp.playerKills);
    }
    w.u32Le(s.civiliansKilled);
    w.u32Le(s.guardsKilled);
    w.u32Le(s.robotsKilled);
    w.u32Le(s.defenseKilled);
    w.u32Le(s.secretsPickedUp);
    w.u32Le(s.secretsReturned);
    w.u32Le(s.secretsStolen);
    w.u32Le(s.secretsDropped);
    w.u32Le(s.powerupsPickedUp);
    w.u32Le(s.deaths);
    w.u32Le(s.kills);
    w.u32Le(s.suicides);
    w.u32Le(s.poisons);
    w.u32Le(s.tractsPlanted);
    w.u32Le(s.grenadesThrown);
    w.u32Le(s.neutronsThrown);
    w.u32Le(s.empsThrown);
    w.u32Le(s.shapedThrown);
    w.u32Le(s.plasmasThrown);
    w.u32Le(s.flaresThrown);
    w.u32Le(s.poisonFlaresThrown);
    w.u32Le(s.healthPacksUsed);
    w.u32Le(s.fixedCannonsPlaced);
    w.u32Le(s.fixedCannonsDestroyed);
    w.u32Le(s.detsPlanted);
    w.u32Le(s.camerasPlanted);
    w.u32Le(s.virusesUsed);
    w.u32Le(s.filesHacked);
    w.u32Le(s.filesReturned);
    w.u32Le(s.creditsEarned);
    w.u32Le(s.creditsSpent);
    w.u32Le(s.healsDone);
    return w.bytes();
}

// ---- Body decoders ------------------------------------------------------

export function decodeVersionReply(r: Reader): VersionResult {
    const ok = r.u8() !== 0;
    let updateUrl = "";
    let sha256: Uint8Array = new Uint8Array(32);
    if (!ok && r.remaining >= 2 + 32) {
        const urlLen = r.u16Le();
        if (urlLen > 0 && urlLen < 512 && r.remaining >= urlLen + 32) {
            updateUrl = utf8Decode(r.bytesN(urlLen));
            sha256 = r.bytesN(32);
        }
    }
    return { ok, updateUrl, sha256 };
}

export function decodeAuthReply(r: Reader): AuthResult {
    const ok = r.u8() !== 0;
    if (ok) {
        return { ok: true, accountId: r.u32Le(), error: "" };
    }
    return { ok: false, accountId: 0, error: r.cstr(256) };
}

export function decodeChatPush(r: Reader): ChatMessage {
    const channel = r.cstr(64);
    const text = r.cstr(MAX_FRAME_PAYLOAD);
    const color = r.u8();
    const brightness = r.u8();
    return { channel, text, color, brightness };
}

export function decodeNewGame(r: Reader): NewGameEvent {
    const status = r.u8();
    const game = decodeLobbyGame(r);
    return { status, game };
}

export const decodeDelGame = (r: Reader): number => r.u32Le();

export const decodeChannel = (r: Reader): string => r.cstr(64);

export function decodeUserInfo(r: Reader): UserInfo {
    const accountId = r.u32Le();
    const agencies = [emptyAgency(), emptyAgency(), emptyAgency(), emptyAgency(), emptyAgency()] as
        UserInfo["agencies"];
    for (let i = 0; i < 5; i++) {
        const a: AgencyStats = {
            wins: r.u16Le(),
            losses: r.u16Le(),
            xpToNextLevel: r.u16Le(),
            level: r.u8(),
            endurance: r.u8(),
            shield: r.u8(),
            jetpack: r.u8(),
            techSlots: r.u8(),
            hacking: r.u8(),
            contacts: r.u8(),
        };
        agencies[i] = a;
    }
    const name = r.lenstr();
    return { accountId, agencies, name };
}

export function decodePresence(r: Reader): PresenceUpdate {
    const action = r.u8();
    const accountId = r.u32Le();
    const gameId = r.u32Le();
    const status = r.u8() as GameStatus;
    const name = r.lenstr();
    return { removed: action === 1, accountId, gameId, status, name };
}

// MOTD: payload is just the body byte if it's a 1-byte terminator,
// else a cstr chunk. payloadSize includes the opcode byte.
export function decodeMotd(r: Reader, payloadSize: number): MotdChunk {
    if (payloadSize <= 2 && r.remaining >= 1) {
        const b = r.u8();
        return { terminator: b === 0, text: b === 0 ? "" : String.fromCharCode(b) };
    }
    return { terminator: false, text: r.cstr(MAX_FRAME_PAYLOAD) };
}

// ---- Symmetric encoders for inbound shapes (used in tests) -------------

export function encodeVersionReply(w: Writer, v: VersionResult): void {
    w.u8(v.ok ? 1 : 0);
    if (!v.ok && v.updateUrl !== "") {
        const urlBytes = utf8Encode(v.updateUrl);
        w.u16Le(urlBytes.length);
        w.write(urlBytes);
        if (v.sha256.length !== 32) throw new CodecError("sha256 must be 32 bytes");
        w.write(v.sha256);
    }
}

export function encodeAuthReply(w: Writer, a: AuthResult): void {
    w.u8(a.ok ? 1 : 0);
    if (a.ok) w.u32Le(a.accountId);
    else w.cstr(a.error);
}

export function encodeChatPush(w: Writer, m: ChatMessage): void {
    w.cstr(m.channel);
    w.cstr(m.text);
    w.u8(m.color);
    w.u8(m.brightness);
}

export function encodeMotdChunk(w: Writer, c: MotdChunk): void {
    if (c.terminator) w.u8(0);
    else w.cstr(c.text);
}

export function encodePresence(w: Writer, p: PresenceUpdate): void {
    w.u8(p.removed ? 1 : 0);
    w.u32Le(p.accountId);
    w.u32Le(p.gameId);
    w.u8(p.status);
    w.lenstr(p.name);
}

export function encodeUserInfoBody(w: Writer, u: UserInfo): void {
    w.u32Le(u.accountId);
    for (const a of u.agencies) {
        w.u16Le(a.wins);
        w.u16Le(a.losses);
        w.u16Le(a.xpToNextLevel);
        w.u8(a.level);
        w.u8(a.endurance);
        w.u8(a.shield);
        w.u8(a.jetpack);
        w.u8(a.techSlots);
        w.u8(a.hacking);
        w.u8(a.contacts);
    }
    w.lenstr(u.name);
}

File: clients/lobby-sdk/cpp/src/codec.cpp
#include "silencer/lobby/codec.h"

#include <cstring>

namespace silencer {
namespace lobby {

// ---- Reader ------------------------------------------------------------

uint8_t Reader::u8() {
    if (off_ >= size_) throw CodecError("u8: short read");
    return data_[off_++];
}

uint16_t Reader::u16_le() {
    if (off_ + 2 > size_) throw CodecError("u16: short read");
    uint16_t v = static_cast<uint16_t>(data_[off_]) |
                 (static_cast<uint16_t>(data_[off_ + 1]) << 8);
    off_ += 2;
    return v;
}

uint32_t Reader::u32_le() {
    if (off_ + 4 > size_) throw CodecError("u32: short read");
    uint32_t v = static_cast<uint32_t>(data_[off_])        |
                 (static_cast<uint32_t>(data_[off_ + 1]) <<  8) |
                 (static_cast<uint32_t>(data_[off_ + 2]) << 16) |
                 (static_cast<uint32_t>(data_[off_ + 3]) << 24);
    off_ += 4;
    return v;
}

void Reader::read_into(uint8_t* dst, size_t n) {
    if (off_ + n > size_) throw CodecError("bytes: short read");
    std::memcpy(dst, data_ + off_, n);
    off_ += n;
}

std::string Reader::cstr(size_t max_len) {
    size_t end = off_;
    size_t limit = off_ + max_len;
    if (limit > size_) limit = size_;
    while (end < limit && data_[end] != 0) ++end;
    if (end >= limit) throw CodecError("cstr: unterminated");
    std::string s(reinterpret_cast<const char*>(data_ + off_), end - off_);
    off_ = end + 1;
    return s;
}

std::string Reader::lenstr() {
    uint8_t n = u8();
    if (off_ + n > size_) throw CodecError("lenstr: short read");
    std::string s(reinterpret_cast<const char*>(data_ + off_), n);
    off_ += n;
    return s;
}

// ---- Writer ------------------------------------------------------------

void Writer::u8(uint8_t v) { buf_.push_back(v); }
void Writer::u16_le(uint16_t v) {
    buf_.push_back(static_cast<uint8_t>(v & 0xFF));
    buf_.push_back(static_cast<uint8_t>((v >> 8) & 0xFF));
}
void Writer::u32_le(uint32_t v) {
    buf_.push_back(static_cast<uint8_t>( v        & 0xFF));
    buf_.push_back(static_cast<uint8_t>((v >>  8) & 0xFF));
    buf_.push_back(static_cast<uint8_t>((v >> 16) & 0xFF));
    buf_.push_back(static_cast<uint8_t>((v >> 24) & 0xFF));
}
void Writer::write(const uint8_t* src, size_t n) {
    buf_.insert(buf_.end(), src, src + n);
}
void Writer::cstr(const std::string& s) {
    buf_.insert(buf_.end(), s.begin(), s.end());
    buf_.push_back(0);
}
void Writer::lenstr(const std::string& s) {
    size_t n = s.size();
    if (n > 255) n = 255;
    buf_.push_back(static_cast<uint8_t>(n));
    buf_.insert(buf_.end(), s.begin(), s.begin() + n);
}

// ---- Framing -----------------------------------------------------------

std::vector<uint8_t> frame_encode(const std::vector<uint8_t>& payload) {
    if (payload.empty() || payload.size() > kMaxFramePayload) {
        throw CodecError("frame_encode: bad payload size");
    }
    std::vector<uint8_t> out;
    out.reserve(1 + payload.size());
    out.push_back(static_cast<uint8_t>(payload.size()));
    out.insert(out.end(), payload.begin(), payload.end());
    return out;
}

bool frame_try_decode(const uint8_t* buf, size_t avail,
                      std::vector<uint8_t>& out_payload, size_t& consumed) {
    if (avail < 1) return false;
    uint8_t n = buf[0];
    if (n == 0) throw CodecError("frame_try_decode: zero-length frame");
    if (avail < static_cast<size_t>(1 + n)) return false;
    out_payload.assign(buf + 1, buf + 1 + n);
    consumed = 1 + n;
    return true;
}

// ---- LobbyGame ---------------------------------------------------------

void encode_lobby_game(Writer& w, const LobbyGame& g) {
    w.u32_le(g.id);
    w.u32_le(g.account_id);
    w.lenstr(g.name);
    w.lenstr(g.password);
    w.lenstr(g.hostname);
    w.lenstr(g.map_name);
    w.write(g.map_hash.data(), g.map_hash.size());
    w.u8(g.players);
    w.u8(g.state);
    w.u8(static_cast<uint8_t>(g.security_level));
    w.u8(g.min_level);
    w.u8(g.max_level);
    w.u8(g.max_players);
    w.u8(g.max_teams);
    w.u8(g.extra);
    w.u16_le(g.port);
}

void decode_lobby_game(Reader& r, LobbyGame& g) {
    g.id         = r.u32_le();
    g.account_id = r.u32_le();
    g.name       = r.lenstr();
    g.password   = r.lenstr();
    g.hostname   = r.lenstr();
    g.map_name   = r.lenstr();
    r.read_into(g.map_hash.data(), g.map_hash.size());
    g.players        = r.u8();
    g.state          = r.u8();
    g.security_level = static_cast<SecurityLevel>(r.u8());
    g.min_level      = r.u8();
    g.max_level      = r.u8();
    g.max_players    = r.u8();
    g.max_teams      = r.u8();
    g.extra          = r.u8();
    g.port           = r.u16_le();
}

// ---- Per-opcode encoders ------------------------------------------------

std::vector<uint8_t> encode_version_request(const std::string& version, Platform p) {
    Writer w;
    w.u8(OpVersion);
    w.cstr(version);
    w.u8(static_cast<uint8_t>(p));
    return std::move(w).take();
}

std::vector<uint8_t> encode_auth_request(const std::string& username,
                                         const std::array<uint8_t, 20>& password_sha1) {
    Writer w;
    w.u8(OpAuth);
    w.cstr(username);
    w.write(password_sha1.data(), password_sha1.size());
    return std::move(w).take();
}

std::vector<uint8_t> encode_chat(const std::string& channel, const std::string& msg) {
    Writer w;
    w.u8(OpChat);
    w.cstr(channel);
    w.cstr(msg);
    return std::move(w).take();
}

std::vector<uint8_t> encode_join_channel(const std::string& current_channel,
                                         const std::string& new_channel) {
    return encode_chat(current_channel, "/join " + new_channel);
}

std::vector<uint8_t> encode_new_game(const LobbyGame& g) {
    Writer w;
    w.u8(OpNewGame);
    encode_lobby_game(w, g);
    return std::move(w).take();
}

std::vector<uint8_t> encode_user_info_request(uint32_t account_id) {
    Writer w;
    w.u8(OpUserInfo);
    w.u32_le(account_id);
    return std::move(w).take();
}

std::vector<uint8_t> encode_ping_ack() {
    Writer w;
    w.u8(OpPing);
    w.u8(1);
    return std::move(w).take();
}

std::vector<uint8_t> encode_upgrade_stat(uint8_t agency_idx, uint8_t stat_id) {
    Writer w;
    w.u8(OpUpgradeStat);
    w.u8(agency_idx);
    w.u8(stat_id);
    return std::move(w).take();
}

std::vector<uint8_t> encode_set_game(uint32_t game_id, GameStatus status) {
    Writer w;
    w.u8(OpSetGame);
    w.u32_le(game_id);
    w.u8(static_cast<uint8_t>(status));
    return std::move(w).take();
}

std::vector<uint8_t> encode_register_stats(uint32_t game_id, uint8_t team_number,
                                           uint32_t account_id, uint8_t stats_agency,
                                           bool won, uint32_t xp,
                                           const MatchStats& s) {
    Writer w;
    w.u8(OpRegisterStats);
    w.u32_le(game_id);
    w.u8(team_number);
    w.u32_le(account_id);
    w.u8(stats_agency);
    w.u8(won ? 1 : 0);
    w.u32_le(xp);
    for (int i = 0; i < 4; ++i) {
        w.u32_le(s.weapons[i].fires);
        w.u32_le(s.weapons[i].hits);
        w.u32_le(s.weapons[i].player_kills);
    }
    w.u32_le(s.civilians_killed);
    w.u32_le(s.guards_killed);
    w.u32_le(s.robots_killed);
    w.u32_le(s.defense_killed);
    w.u32_le(s.secrets_picked_up);
    w.u32_le(s.secrets_returned);
    w.u32_le(s.secrets_stolen);
    w.u32_le(s.secrets_dropped);
    w.u32_le(s.powerups_picked_up);
    w.u32_le(s.deaths);
    w.u32_le(s.kills);
    w.u32_le(s.suicides);
    w.u32_le(s.poisons);
    w.u32_le(s.tracts_planted);
    w.u32_le(s.grenades_thrown);
    w.u32_le(s.neutrons_thrown);
    w.u32_le(s.emps_thrown);
    w.u32_le(s.shaped_thrown);
    w.u32_le(s.plasmas_thrown);
    w.u32_le(s.flares_thrown);
    w.u32_le(s.poison_flares_thrown);
    w.u32_le(s.health_packs_used);
    w.u32_le(s.fixed_cannons_placed);
    w.u32_le(s.fixed_cannons_destroyed);
    w.u32_le(s.dets_planted);
    w.u32_le(s.cameras_planted);
    w.u32_le(s.viruses_used);
    w.u32_le(s.files_hacked);
    w.u32_le(s.files_returned);
    w.u32_le(s.credits_earned);
    w.u32_le(s.credits_spent);
    w.u32_le(s.heals_done);
    return std::move(w).take();
}

// ---- Body decoders ------------------------------------------------------

VersionResult decode_version_reply(Reader& r) {
    VersionResult v;
    v.ok = r.u8() != 0;
    if (!v.ok && r.remaining() >= 2 + 32) {
        uint16_t url_len = r.u16_le();
        if (url_len > 0 && url_len < 512 && r.remaining() >= static_cast<size_t>(url_len) + 32) {
            std::string url(url_len, '\0');
            r.read_into(reinterpret_cast<uint8_t*>(&url[0]), url_len);
            v.update_url = std::move(url);
            r.read_into(v.sha256.data(), v.sha256.size());
        }
    }
    return v;
}

AuthResult decode_auth_reply(Reader& r) {
    AuthResult a;
    a.ok = r.u8() != 0;
    if (a.ok) {
        a.account_id = r.u32_le();
    } else {
        a.error = r.cstr(256);
    }
    return a;
}

ChatMessage decode_chat_push(Reader& r) {
    ChatMessage m;
    m.channel    = r.cstr(64);
    m.text       = r.cstr(kMaxFramePayload);
    m.color      = r.u8();
    m.brightness = r.u8();
    return m;
}

NewGameEvent decode_new_game(Reader& r) {
    NewGameEvent ev;
    ev.status = r.u8();
    decode_lobby_game(r, ev.game);
    return ev;
}

uint32_t decode_del_game(Reader& r) { return r.u32_le(); }

std::string decode_channel(Reader& r) { return r.cstr(64); }

UserInfo decode_user_info(Reader& r) {
    UserInfo u;
    u.account_id = r.u32_le();
    for (int i = 0; i < 5; ++i) {
        AgencyStats& a = u.agencies[i];
        a.wins             = r.u16_le();
        a.losses           = r.u16_le();
        a.xp_to_next_level = r.u16_le();
        a.level            = r.u8();
        a.endurance        = r.u8();
        a.shield           = r.u8();
        a.jetpack          = r.u8();
        a.tech_slots       = r.u8();
        a.hacking          = r.u8();
        a.contacts         = r.u8();
    }
    u.name = r.lenstr();
    return u;
}

PresenceUpdate decode_presence(Reader& r) {
    PresenceUpdate p;
    uint8_t action = r.u8();
    p.removed     = (action == 1);
    p.account_id  = r.u32_le();
    p.game_id     = r.u32_le();
    p.status      = static_cast<GameStatus>(r.u8());
    p.name        = r.lenstr();
    return p;
}

MotdChunk decode_motd(Reader& r, size_t payload_size) {
    // payload_size includes the opcode byte; if the body is just a
    // single 0x00 byte, that's the terminator.
    MotdChunk c;
    if (payload_size <= 2 && r.remaining() >= 1) {
        uint8_t b = r.u8();
        c.terminator = (b == 0);
        if (!c.terminator) c.text.push_back(static_cast<char>(b));
        return c;
    }
    c.terminator = false;
    c.text       = r.cstr(kMaxFramePayload);
    return c;
}

// ---- Symmetric encoders for inbound shapes (used in tests) -------------

void encode_user_info_body(Writer& w, const UserInfo& u) {
    w.u32_le(u.account_id);
    for (int i = 0; i < 5; ++i) {
        const AgencyStats& a = u.agencies[i];
        w.u16_le(a.wins);
        w.u16_le(a.losses);
        w.u16_le(a.xp_to_next_level);
        w.u8(a.level);
        w.u8(a.endurance);
        w.u8(a.shield);
        w.u8(a.jetpack);
        w.u8(a.tech_slots);
        w.u8(a.hacking);
        w.u8(a.contacts);
    }
    w.lenstr(u.name);
}

void encode_auth_reply(Writer& w, const AuthResult& a) {
    w.u8(a.ok ? 1 : 0);
    if (a.ok) {
        w.u32_le(a.account_id);
    } else {
        w.cstr(a.error);
    }
}

void encode_version_reply(Writer& w, const VersionResult& v) {
    w.u8(v.ok ? 1 : 0);
    if (!v.ok && !v.update_url.empty()) {
        w.u16_le(static_cast<uint16_t>(v.update_url.size()));
        w.write(reinterpret_cast<const uint8_t*>(v.update_url.data()),
                v.update_url.size());
        w.write(v.sha256.data(), v.sha256.size());
    }
}

void encode_chat_push(Writer& w, const ChatMessage& m) {
    w.cstr(m.channel);
    w.cstr(m.text);
    w.u8(m.color);
    w.u8(m.brightness);
}

void encode_motd_chunk(Writer& w, const MotdChunk& c) {
    if (c.terminator) {
        w.u8(0);
    } else {
        w.cstr(c.text);
    }
}

void encode_presence(Writer& w, const PresenceUpdate& p) {
    w.u8(p.removed ? 1 : 0);
    w.u32_le(p.account_id);
    w.u32_le(p.game_id);
    w.u8(static_cast<uint8_t>(p.status));
    w.lenstr(p.name);
}

} // namespace lobby
} // namespace silencer
